File: 03_JS/3.js
/*
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?
*/

function isPrime(n) {
  const wn = Math.sqrt(n);
  for(let i = 2; i <= wn; i++) {
    if(n % i == 0) {
      return false;
      break;
    }
  }
  return true;
}

function smallestPrimeDivisor(n) {
  let current = 2;
  while(true) {
    if(isPrime(current) && n % current == 0) {
      return current;
    }
    current++;
  }
}

function primeFactors(n) {
  if(isPrime(n)) {
    return [n];
  } else {
    const p = smallestPrimeDivisor(n);
    return  [p].concat(primeFactors(n/p));
  }
}

const factors = primeFactors(600851475143);

console.log(Math.max(...factors));
//6857

File: 03_PY/3.py
#!/usr/bin/python

#The prime factors of 13195 are 5, 7, 13 and 29.
#What is the largest prime factor of the number 600851475143 ?

import math

def isPrime(n):
  if n==1:
    return False
  rn = int(math.ceil(math.sqrt(n)))
  for x in xrange(2, rn):
    if n%x == 0:
      return False
  return True

def smallestPrimeFactor(n):
  for x in xrange(2, n+1):
    if isPrime(x) and n%x == 0:
      return x

def allPrimeFactors(n):
  p = smallestPrimeFactor(n)
  nn = n/p
  if nn == 1:
    return [p]
  else:
    return [p] + allPrimeFactors(nn)

print allPrimeFactors(600851475143)[-1]